**4.8 Grouping Data & Aggregating Variables - Task **


Importing Libraries

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

Import Data

In [2]:
#Create path
path = r'C:\Users\mngit\OneDrive\Data Analytics\Project Excercises\Achievement 4\Instacart Basket Analysis' 

In [3]:
# import ords_prods_merged
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))

In [4]:
#find aggregated mean of the “order_number” column grouped by “department_id”
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

Difference between subset and dataframe aggregated mean results:
 i)The subset only included 8 departments while the entire dataframe includes all 21 departments.
ii)The mean value is different for each department, for instance 'dept_id 4' had a mean value of 18.83 in the subset but has a   value of 17.81 in the dataframe, 'dept_id 17' has a value of 11.29 in the subset but has a value of 15.69 in dataframe

In [5]:
#create new column containing the maximum frequency of the “order_number” 
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [6]:
ords_prods_merge.head(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   
5   3367565        1    prior             6                   2   
6    550135        1    prior             7                   1   
7   3108588        1    prior             8                   1   
8   2295261        1    prior             9                   1   
9   2550362        1    prior            10                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     7.0         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   
5                  7                    19.0         196                  1   
6                  9                    20.0         196                  1   
7                 14                    14.0         196                  2   
8                 16                     0.0         196                  4   
9                  8                    30.0         196                  1   

   reordered _merge product_name  aisle_id  department_id  prices  \
0          0   both         Soda        77              7     9.0   
1          1   both         Soda        77              7     9.0   
2          1   both         Soda        77              7     9.0   
3          1   both         Soda        77              7     9.0   
4          1   both         Soda        77              7     9.0   
5          1   both         Soda        77              7     9.0   
6          1   both         Soda        77              7     9.0   
7          1   both         Soda        77              7     9.0   
8          1   both         Soda        77              7     9.0   
9          1   both         Soda        77              7     9.0   

     price_range_loc     busiest_day     busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy   Regularly busy        Average orders   
1  Mid-range product  Regularly busy  Least busy days        Average orders   
2  Mid-range product  Regularly busy  Least busy days           Most orders   
3  Mid-range product      Least busy  Least busy days        Average orders   
4  Mid-range product      Least busy  Least busy days           Most orders   
5  Mid-range product  Regularly busy   Regularly busy        Average orders   
6  Mid-range product  Regularly busy     Busiest days         Fewest orders   
7  Mid-range product  Regularly busy     Busiest days           Most orders   
8  Mid-range product  Regularly busy     Busiest days           Most orders   
9  Mid-range product      Least busy  Least busy days        Average orders   

   max_order  
0         10  
1         10  
2         10  
3         10  
4         10  
5         10  
6         10  
7         10  
8         10  
9         10

In [7]:
ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
..       ...      ...      ...           ...                 ...   
95   3226575      360    prior             1                   5   
96   1469869      377    prior             3                   5   
97   1927023      387    prior             2                   4   
98    858092      420    prior             4                   1   
99    119314      420    prior             5                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     7.0         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
..                ...                     ...         ...                ...   
95                 12                     7.0         196                  1   
96                 17                     3.0         196                  9   
97                 10                    22.0         196                  3   
98                 19                    30.0         196                  2   
99                 19                    16.0         196                  2   

    reordered _merge product_name  aisle_id  department_id  prices  \
0           0   both         Soda        77              7     9.0   
1           1   both         Soda        77              7     9.0   
2           1   both         Soda        77              7     9.0   
3           1   both         Soda        77              7     9.0   
4           1   both         Soda        77              7     9.0   
..        ...    ...          ...       ...            ...     ...   
95          0   both         Soda        77              7     9.0   
96          0   both         Soda        77              7     9.0   
97          0   both         Soda        77              7     9.0   
98          0   both         Soda        77              7     9.0   
99          1   both         Soda        77              7     9.0   

      price_range_loc     busiest_day     busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy   Regularly busy        Average orders   
1   Mid-range product  Regularly busy  Least busy days        Average orders   
2   Mid-range product  Regularly busy  Least busy days           Most orders   
3   Mid-range product      Least busy  Least busy days        Average orders   
4   Mid-range product      Least busy  Least busy days           Most orders   
..                ...             ...              ...                   ...   
95  Mid-range product  Regularly busy   Regularly busy           Most orders   
96  Mid-range product  Regularly busy   Regularly busy           Most orders   
97  Mid-range product      Least busy  Least busy days           Most orders   
98  Mid-range product  Regularly busy     Busiest days        Average orders   
99  Mid-range product  Regularly busy  Least busy days        Average orders   

    max_order  
0          10  
1          10  
2          10  
3          10  
4          10  
..        ...  
95          3  
96          3  
97          8  
98         22  
99         22  

[100 rows x 20 columns]

Create column containing a flag designating whether a customer is “loyal”, "Regular" or "New"

In [8]:
#Create 'Loyal customer'column
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [9]:
#Create 'Regular customer'column
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [10]:
#Create 'New customer'column
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [11]:
#print frequency of “loyalty_flag” 
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [12]:
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(40)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3

In [13]:
# Aggregating mean price by customer loyalty type
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.386336
New customer      13.294670
Regular customer  12.495717

Spending Habits:
New customers spent on average 13.2
Regular customers spent on average 12.5
Loyal customers spent 10.4

Create column containing types of spenders flag

In [14]:
#create 'mean_prices' column
ords_prods_merge['mean_prices'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [15]:
ords_prods_merge.head(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   
5   3367565        1    prior             6                   2   
6    550135        1    prior             7                   1   
7   3108588        1    prior             8                   1   
8   2295261        1    prior             9                   1   
9   2550362        1    prior            10                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     7.0         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   
5                  7                    19.0         196                  1   
6                  9                    20.0         196                  1   
7                 14                    14.0         196                  2   
8                 16                     0.0         196                  4   
9                  8                    30.0         196                  1   

   reordered  ... aisle_id department_id  prices    price_range_loc  \
0          0  ...       77             7     9.0  Mid-range product   
1          1  ...       77             7     9.0  Mid-range product   
2          1  ...       77             7     9.0  Mid-range product   
3          1  ...       77             7     9.0  Mid-range product   
4          1  ...       77             7     9.0  Mid-range product   
5          1  ...       77             7     9.0  Mid-range product   
6          1  ...       77             7     9.0  Mid-range product   
7          1  ...       77             7     9.0  Mid-range product   
8          1  ...       77             7     9.0  Mid-range product   
9          1  ...       77             7     9.0  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day max_order  \
0  Regularly busy   Regularly busy        Average orders        10   
1  Regularly busy  Least busy days        Average orders        10   
2  Regularly busy  Least busy days           Most orders        10   
3      Least busy  Least busy days        Average orders        10   
4      Least busy  Least busy days           Most orders        10   
5  Regularly busy   Regularly busy        Average orders        10   
6  Regularly busy     Busiest days         Fewest orders        10   
7  Regularly busy     Busiest days           Most orders        10   
8  Regularly busy     Busiest days           Most orders        10   
9      Least busy  Least busy days        Average orders        10   

   loyalty_flag  mean_prices  
0  New customer     6.367797  
1  New customer     6.367797  
2  New customer     6.367797  
3  New customer     6.367797  
4  New customer     6.367797  
5  New customer     6.367797  
6  New customer     6.367797  
7  New customer     6.367797  
8  New customer     6.367797  
9  New customer     6.367797  

[10 rows x 22 columns]

Creating spending_flag

In [16]:
#Create 'Low spender'flag
ords_prods_merge.loc[ords_prods_merge['mean_prices'] < 10, 'spending_flag'] = 'Low spender'

In [17]:
#Create 'High spender'flag
ords_prods_merge.loc[ords_prods_merge['mean_prices'] >= 10, 'spending_flag'] = 'High spender'

In [18]:
#print frequency of “spending_flag” 
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

Create column containing order frequency flag

In [19]:
#create 'median_days_since_prior_order' column
ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [20]:
ords_prods_merge.head(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   
5   3367565        1    prior             6                   2   
6    550135        1    prior             7                   1   
7   3108588        1    prior             8                   1   
8   2295261        1    prior             9                   1   
9   2550362        1    prior            10                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     7.0         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   
5                  7                    19.0         196                  1   
6                  9                    20.0         196                  1   
7                 14                    14.0         196                  2   
8                 16                     0.0         196                  4   
9                  8                    30.0         196                  1   

   reordered  ... prices    price_range_loc     busiest_day     busiest_days  \
0          0  ...    9.0  Mid-range product  Regularly busy   Regularly busy   
1          1  ...    9.0  Mid-range product  Regularly busy  Least busy days   
2          1  ...    9.0  Mid-range product  Regularly busy  Least busy days   
3          1  ...    9.0  Mid-range product      Least busy  Least busy days   
4          1  ...    9.0  Mid-range product      Least busy  Least busy days   
5          1  ...    9.0  Mid-range product  Regularly busy   Regularly busy   
6          1  ...    9.0  Mid-range product  Regularly busy     Busiest days   
7          1  ...    9.0  Mid-range product  Regularly busy     Busiest days   
8          1  ...    9.0  Mid-range product  Regularly busy     Busiest days   
9          1  ...    9.0  Mid-range product      Least busy  Least busy days   

   busiest_period_of_day max_order  loyalty_flag mean_prices spending_flag  \
0         Average orders        10  New customer    6.367797   Low spender   
1         Average orders        10  New customer    6.367797   Low spender   
2            Most orders        10  New customer    6.367797   Low spender   
3         Average orders        10  New customer    6.367797   Low spender   
4            Most orders        10  New customer    6.367797   Low spender   
5         Average orders        10  New customer    6.367797   Low spender   
6          Fewest orders        10  New customer    6.367797   Low spender   
7            Most orders        10  New customer    6.367797   Low spender   
8            Most orders        10  New customer    6.367797   Low spender   
9         Average orders        10  New customer    6.367797   Low spender   

   median_days_since_prior_order  
0                           20.0  
1                           20.0  
2                           20.0  
3                           20.0  
4                           20.0  
5                           20.0  
6                           20.0  
7                           20.0  
8                           20.0  
9                           20.0  

[10 rows x 24 columns]

Creating order frequency flag

In [21]:
#Create 'Non-frequent customer'flag
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [22]:
#Create 'Regular customer'flag
ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order'] > 10) & (ords_prods_merge['median_days_since_prior_order'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [23]:
#Create 'Frequent customer'flag
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] <=10, 'order_frequency_flag'] = 'Frequent customer'

In [24]:
#print frequency of “order frequency flag” 
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        22790523
Regular customer          6927608
Non-frequent customer     2686728
Name: order_frequency_flag, dtype: int64

In [25]:
ords_prods_merge.head(60)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   2968173       15    prior            15                   1   
11   1870022       15    prior            17                   2   
12   1911383       15    prior            18                   2   
13   2715276       15    prior            21                   1   
14    487368       15    prior            22                   1   
15   2293453       19    prior             2                   5   
16   1973799       19    prior             5                   6   
17    532817       19    prior             7                   4   
18   1573906       21    prior            10                   3   
19   1593000       31    prior            10                   3   
20   2231262       31    prior            17                   3   
21   2580647       43    prior             6                   4   
22   2187180       43    prior             9                   4   
23   2497897       52    prior             1                   1   
24   1318871       52    prior             2                   1   
25   1261384       52    prior             3                   1   
26    580568       52    prior             5                   2   
27   2428073       52    prior             7                   2   
28    342306       52    prior             9                   1   
29   1520353       52    prior            10                   2   
30   2542086       52    prior            11                   2   
31    944694       52    prior            12                   1   
32   1268191       52    prior            15                   2   
33   1122089       52    prior            17                   2   
34   1498922       52    prior            18                   2   
35    180919       52    prior            19                   3   
36   2695875       52    prior            21                   5   
37   3205960       67    prior             1                   3   
38    970922       67    prior             2                   1   
39    596585       67    prior             3                   1   
40   2037087       67    prior             4                   2   
41   2474111       67    prior             5                   3   
42   1568481       67    prior             6                   3   
43   2048527       67    prior             8                   2   
44   2389031       67    prior             9                   2   
45   2378562       67    prior            10                   1   
46   1515698       67    prior            11                   1   
47   1928050       67    prior            12                   2   
48   1949274       67    prior            13                   1   
49    551869       67    prior            14                   1   
50    768572       67    prior            16                   4   
51    150528       67    prior            18                   5   
52     29758       67    prior            19                   0   
53   1558976       67    prior            21                   1   
54    863974       67    prior            22                   1   
55   2308399       67    prior            23                   1   
56   2759128       81    prior             3                   1   
57   1812985       81    prior             5                

In [26]:
# Export data to Pickle
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))

In [27]:
#replaced all your outliers with missing values, your max value is a much more realistic price point of 25.
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan